Add explanations (#32)
canonical/ubuntu-packaging-guide: docs/explanation/launchpad.rst

                |------------+----------+-------------|
                | Lorem      | ipsum    | dolor       |
                | sit        | amet     | consectetur |
                | adipiscing | elit     | sed         |

          - | \| Column 1 \| Column 2 \| Column 3 \|  
            | \|\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-\|  
            | \| Lorem | ipsum | dolor \|  
            | \| sit \| amet \| consectetur \|  
            | \| adipiscing \| elit \| sed \|

    or long chunks of whitespace between two sections:

    .. list-table::
        :header-rows: 1

        * - Input
          - Output
        * - .. code:: text

                Here are two paragraphs with lots   
                of whitespace between them.
                
                
                
                
                But they're still just two paragraphs

          - | Here are two paragraphs with lots of whitespace between them.
            | 
            | But they're still just two paragraphs

Getting help
------------

If you need help with Launchpad you can choose any of the following methods:

IRC chatrooms
~~~~~~~~~~~~~

> On the ``irc.libera.chat`` :term:`IRC` server you will find the ``#launchpad`` channel, where you

Reviewer: It would be helpful for readers if we actually linked directly to this channel on Libera

Author: Do you mean like this?

```rst
`#launchpad <irc://irc.libera.chat/launchpad>`_
```

Or do you mean to move it to the start of the article?

Reviewer: I mean exactly like that! :) So they can just one-click to go to the channel

Comment Markup
--------------

Unfortunately Launchpad only regognizes a very limited set of markup patterns when you write comments. 

.. note::

    
    Support for a wider range of markup patterns is a very comman and old request/wish; 
    take for example LP: `#391780 <https://bugs.launchpad.net/launchpad/+bug/391780>`_.
    
    You can "upvote" (mark yourself as effected) or leave a comment to this bug report to
    show your support for the feature request.

    **Reminder:** Please stay civil! The Launchpad team has only limited resources.

The full range of pattern is documented `here <https://help.launchpad.net/Comments>`_. 
The most important patterns are:

URIs
~~~~

Launchpad can recognize ``http``, ``https``, ``ftp``, ``sftp``, ``mailto``,
``news``, ``irc`` and ``jabber`` :term:`URI's <URI>`.

.. note::

    ``tel``, ``urn``, ``telnet``, ``ldap`` :term:`URI's <URI>`, relative 
    :term:`URL's <URL>` like ``example.com`` and E-Mail addresses like 
    ``test@example.com`` are **NOT** recognized. 

Referencing Launchpad Bugs
~~~~~~~~~~~~~~~~~~~~~~~~~~

Synopsis
^^^^^^^^

> 

Reviewer: I got a little confused here, I'm not sure synopsis is very descriptive as a title. It might help to have a little sentence between the title and the code block to explain it a bit?

Reviewer: I'm also a bit confused by the table and what its purpose is - might be worth just clarifying that with a short introduction as well

Author: > I got a little confused here, I'm not sure synopsis is very descriptive as a title. It might help to have a little sentence between the title and the code block to explain it a bit?

I thought this is an established concept; like in this man page: https://manpages.ubuntu.com/manpages/lunar/en/man1/bash.1.html

But I agree that this can be confusing for someone not aware of this :/ There will be more sections like this where the synopsis will just show the syntax e.g. of a command. Do you have a general recommendation what to write there?

Author: > I'm also a bit confused by the table and what its purpose is - might be worth just clarifying that with a short introduction as well

Good Point! The input is what you type in a launchpad text field and the output shows how it will be rendered on Launchpad. I add a short description to explain this.

Reviewer: > > I got a little confused here, I'm not sure synopsis is very descriptive as a title. It might help to have a little sentence between the title and the code block to explain it a bit?
> 
> I thought this is an established concept; like in this man page: https://manpages.ubuntu.com/manpages/lunar/en/man1/bash.1.html
> 
> But I agree that this can be confusing for someone not aware of this :/ There will be more sections like this where the synopsis will just show the syntax e.g. of a command. Do you have a general recommendation what to write there?

If it's the expected terminology then stick with it, since the typical user of the packaging guide can be expected to know it. Let's chalk it up to "a thing I just needed to learn" :)

Author: I changed "Synopsis" to "Pattern" and moved the entire Launchpad comment section into a reference article as the section is barely explanatory content.

Resolved with 3d3ffc9a8b2db303ff33bbc35872d7d980f8ca3e

.. code:: text

    LP: #<LP-Bug-Number>[, #<LP-Bug-Number>]...

.. note::

>     This pattern is case invariant. The amount of :term:`Whitespace` can be 
>     variable, but if you place :term:`Whitespace` anywhere else; the 

Reviewer: ```suggestion
    This pattern is case invariant. The amount of :term:`whitespace <Whitespace>` can be 
    variable, but if you place whitespace anywhere else the 
```

>     :term:`Regular Expression` might not parse the input correctly. 

Reviewer: ```suggestion
    :term:`regular expression <Regular Expression>` might not parse the input correctly. 
```

            #2,
            #3,
            #4
            lp: #1
            (lp: #1)
            lp: #1, #2.
            LP #1
            LP: #1 , #2
            LP: #1, #2,

            #3

      - | LP: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_
        | (LP: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_)
        | LP: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_, `#2 <https://bugs.launchpad.net/ubuntu/+bug/2>`_.
        | LP:
        | `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_,
        | `#2 <https://bugs.launchpad.net/ubuntu/+bug/2>`_,
        | `#3 <https://bugs.launchpad.net/ubuntu/+bug/3>`_,
        | `#4 <https://bugs.launchpad.net/ubuntu/+bug/4>`_
        | lp: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_
        | (lp: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_)
        | lp: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_, `#2 <https://bugs.launchpad.net/ubuntu/+bug/2>`_.
        | LP #1
        | LP: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_ , #2
        | LP: `#1 <https://bugs.launchpad.net/ubuntu/+bug/1>`_, `#2 <https://bugs.launchpad.net/ubuntu/+bug/2>`_,
        |
        | #3 

Whitespaces
~~~~~~~~~~~

Launchpad will

> - cut off any :term:`Whitespace` to the right and
> - keep any :term:`Whitespace` to the left and
> - reduce any :term:`Whitespace` between non-:term:`Whitespace` characters to 

Reviewer: ```suggestion
- cut off any whitespace to the right and
- keep any whitespace to the left and
- reduce any whitespace between non-whitespace characters to 
```

  just one (this includes new-line characters as well).

.. note::

>     Technically Launchpad passes :term:`Whitespace` through and 
>     the browser just ignores the :term:`Whitespace`.

Reviewer: ```suggestion
    Technically, Launchpad passes whitespace through and 
    the browser just ignores the whitespace.
```

.. warning::

    Because of the above described behaviour you will have a bad time trying to write a table:

    .. list-table::
        :header-rows: 1

        * - Input
          - Output
        * - .. code:: text

                | Column 1   | Column 2 | Column 3    |
                |------------+----------+-------------|
                | Lorem      | ipsum    | dolor       |
                | sit        | amet     | consectetur |
                | adipiscing | elit     | sed         |

          - | \| Column 1 \| Column 2 \| Column 3 \|  
            | \|\-\-\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-+\-\-\-\-\-\-\-\-\-\-\-\-\-\|  
            | \| Lorem | ipsum | dolor \|  
            | \| sit \| amet \| consectetur \|  
            | \| adipiscing \| elit \| sed \|

>     or long chunks of :term:`Whitespace` between two sections:

Reviewer: ```suggestion
    or long chunks of whitespace between two sections:
```

    .. list-table::
        :header-rows: 1

        * - Input
          - Output
        * - .. code:: text

                Here are two paragraphs with lots   
                of whitespace between them.
                
                
                
                
                But they're still just two paragraphs

          - | Here are two paragraphs with lots of whitespace between them.
            | 
            | But they're still just two paragraphs

> Getting Help

Reviewer: ```suggestion
Getting help
```

------------

If you need help with Launchpad you can choose any of the following methods:

> IRC Chatrooms

Reviewer: ```suggestion
IRC chatrooms
```

~~~~~~~~~~~~~

> On the ``irc.libera.chat`` :term:`IRC` server exists the ``#launchpad`` channel, where you
> can ask the Launchpad team and the :term:`Ubuntu` community for help.

Reviewer: ```suggestion
On the ``irc.libera.chat`` :term:`IRC` server you will find the ``#launchpad`` channel, where you
can ask the Launchpad team and the Ubuntu community for help.
```

> Mailing Lists

Reviewer: ```suggestion
Mailing lists
```

~~~~~~~~~~~~~

> If you prefer to ask for help via E-Mail, you can write to the

Reviewer: ```suggestion
If you prefer to ask for help via email, you can write to the
```

`launchpad-users <https://launchpad.net/~launchpad-users>`_ 
mailing list (``launchpad-users@lists.launchpad.net``).

Ask a question
~~~~~~~~~~~~~~

> As mentioned above, Launchpad has a community FAQ feature (called "Answers") where
> you can see other people's questions or ask one yourself. Use can use the *Answers*
> feature of the Launchpad project on Launchpad itself: https://answers.launchpad.net/launchpad

Reviewer: ```suggestion
As mentioned above, Launchpad has a `community FAQ feature <https://answers.launchpad.net/launchpad>`_ called "Answers" where
you can see other people's questions or ask one yourself.
```

Report a bug
~~~~~~~~~~~~

> If you encounter any bug related to Launchpad, you can submit a bug report to the
> :term:`Bug Tracking System` of the Launchpad project on Launchpad itself: 
> https://bugs.launchpad.net/launchpad

Reviewer: ```suggestion
If you encounter any bug related to Launchpad, you can submit a bug report to the
:term:`Bug Tracking System` of the Launchpad project `on Launchpad itself <https://bugs.launchpad.net/launchpad>`_.
```

> Staging Environment

Reviewer: ```suggestion
Staging environment
```

-------------------

Before new features are deployed to the production environment they get

> deployed to a staging environment (https://qastaging.launchpad.net/) where

Reviewer: ```suggestion
`deployed to a staging environment <https://qastaging.launchpad.net/>`_ where
```

the changes can get tested.

> You can use the staging environment, to try out :term:`Launchpad` features.

Reviewer: ```suggestion
You can use the staging environment, to try out Launchpad features.
```

API
---

Launchpad has a Web :term:`API` that you can use to interact with its services.

> This makes it easy for developer communities like :term:`Ubuntu's <Ubuntu>` to

Reviewer: ```suggestion
This makes it easy for developer communities like Ubuntu's to
```

automate specific workflows.

> You can find the reference documentation for the Web :term:`API`:
> https://launchpad.net/+apidoc/

Reviewer: ```suggestion
You can find the reference `documentation for the web API <https://launchpad.net/+apidoc/>`_ on Launchpad.
```

The Launchpad team even created an :term:`Open Source <Open Source Software>`

> ython library ``launchpadlib``: https://help.launchpad.net/API/launchpadlib

Reviewer: ```suggestion
Python library, `launchpadlib <https://help.launchpad.net/API/launchpadlib>`_
```

Resources
---------

> - `Launchpad Home Page <Launchpad_>`_
> - `The Launchpad software project on Launchpad itself <https://launchpad.net/launchpad>`_
>     - `Launchpad Bug Tracker <https://bugs.launchpad.net/launchpad>`_
>     - `Launchpad Questions & Answers <https://answers.launchpad.net/launchpad>`_
> - `Launchpad Wiki <https://help.launchpad.net/>`_
> - `Launchpad Development Wiki <https://dev.launchpad.net/>`_
> - `Launchpad Blog <https://blog.launchpad.net/>`_
> - :doc:`/reference/git-ubuntu`

Reviewer: ```suggestion
- `Launchpad home page <Launchpad_>`_
- `The Launchpad software project on Launchpad itself <https://launchpad.net/launchpad>`_
    - `Launchpad bug tracker <https://bugs.launchpad.net/launchpad>`_
    - `Launchpad questions and answers <https://answers.launchpad.net/launchpad>`_
- `Launchpad Wiki <https://help.launchpad.net/>`_
- `Launchpad Development Wiki <https://dev.launchpad.net/>`_
- `Launchpad blog <https://blog.launchpad.net/>`_
- :doc:`/reference/git-ubuntu`
```
